src/lib/metrics.rs

```diff
@@ -204,6 +204,94 @@ pub struct DemuxedGroupMetrics<'a> {
+    // the "best" barcode.
+    let best_barcode_count = per_group_metrics[0..per_group_metrics.len() - 1]
+        .iter()
+        .max_by_key(|s| s.total_matches)
+        .map_or(0, |s| s.total_matches);
+
+    // Build per group metrics to output.  We need to derive (e.g. ratio, percentage) values here.
+    let per_group_metrics: Vec<SampleMetricsProcessed> = samples
+        .iter()
+        .zip(per_group_metrics.iter())
+        .map(|(sample, metrics)| metrics.as_processed(total_templates, best_barcode_count, sample))
+        .collect();
+
+    // Finally output it to a file
+    let output_path = output_dir.as_ref().join(filename);
+    let delim = DelimFile::default();
+    delim.write_tsv(&output_path, per_group_metrics)?;
+
+    Ok(())
+}
+
+/// Builds a new set of "samples" and demux metrics by grouping input samples by project.  Each
+/// new sample corresponds to a unique project, and the associated metric are the sum of values
+/// for samples with the same project.  The undetermined sample is not grouped with any other
+/// sample.
+fn build_per_project_metrics(
+    samples: &[SampleMetadata],
+    per_sample_metrics: &[DemuxedGroupSampleMetrics],
+) -> Result<(Vec<SampleMetadata>, Vec<DemuxedGroupSampleMetrics>)> {
+    let mut project_to_metric: AHashMap<Option<&String>, DemuxedGroupSampleMetrics> =
+        AHashMap::new();
+    let mut project_to_sample: AHashMap<Option<&String>, SampleMetadata> = AHashMap::new();
+    let num_per_group_metrics = per_sample_metrics.len();
+
+    // Iterate over all samples, except the undetermined sample
+    for (sample, metric) in
+        samples.iter().zip(per_sample_metrics.iter()).take(num_per_group_metrics - 1)
+    {
+        let key = sample.project.as_ref();
+        match project_to_metric.get_mut(&key) {
+            None => {
+                // Build a new dummy sample for this project
+                let project = sample.project.clone().unwrap_or_else(|| String::from("None"));
+                let sample = SampleMetadata::new_allow_invalid_bases(
+                    project,
+                    BString::from(vec![b'N'; samples[0].barcode.len()]),
+                    project_to_metric.len(),
+                )?;
+                // Add the sample and current metric to the maps
+                project_to_sample.insert(key, sample);
+                project_to_metric.insert(key, metric.clone());
+            }
+            Some(project_metric) => project_metric.update_with_self(metric.clone()),
+        };
+    }
+
+    let mut project_metrics: Vec<DemuxedGroupSampleMetrics> =
+        Vec::with_capacity(project_to_metric.len() + 1);
+    let mut project_samples: Vec<SampleMetadata> = Vec::with_capacity(project_to_metric.len() + 1);
+
+    // Sort by project name
+    for key in project_to_metric.keys().sorted() {
+        let metric = project_to_metric.get(key).unwrap();
+        let sample = project_to_sample.get(key).unwrap();
```

Comment:
Could use `map.remove()` here to take the element out and take ownership so you don't have to clone() below.